### PR TITLE
Refactor `StringifyTypeExpr`

### PR DESCRIPTION
**Pull request by josh11b**
* Make the step stack into a class
* Make the operations on the stack (pushing, popping, test for done) into methods on the stack class.
* Add more kinds of steps (array bound and name).
* Rewrite cases to use the new kinds of steps. Afterward, none use `Step::Next()` or the step index, so those get removed.
* Add another convenience method `PushTypeId`.
* Remove the `SemIR::File&` member from the steps, since it doesn't change.

Hopefully using `step_stack.Push`... calls makes it clear that they are resolved in the reverse order they are executed.

### CODE REVIEW

File: toolchain/sem_ir/stringify_type.cpp [hunk 1]
```diff
@@ -24,78 +24,60 @@ static auto GetTypePrecedence(InstKind kind) -> int {
...
-auto StringifyTypeExpr(const SemIR::File& outer_sem_ir, InstId outer_inst_id)
+auto StringifyTypeExpr(const SemIR::File& sem_ir, InstId outer_inst_id)
     -> std::string {
   std::string str;
   llvm::raw_string_ostream out(str);
 
-  struct Step {
-    // The instruction's file.
-    const File& sem_ir;
+  class StepStack {
```
Comment (zygoloid) [inline toolchain/sem_ir/stringify_type.cpp]: Would it make sense to move this out of the function? That might enable more refactoring, such as splitting this function into smaller functions, at a later stage.

File: toolchain/sem_ir/stringify_type.cpp [hunk 2]
```diff
@@ -135,21 +148,19 @@ auto StringifyTypeExpr(const SemIR::File& outer_sem_ir, InstId outer_inst_id)
...
-          steps.push_back(step.Next());
-          push_inst_id(sem_ir.types().GetInstId(inst.element_type_id));
-        } else if (step.index == 1) {
-          out << "; " << sem_ir.GetArrayBoundValue(inst.bound_id) << "]";
-        }
+        out << "[";
+        step_stack.PushString("]");
+        step_stack.PushArrayBound(inst.bound_id);
+        step_stack.PushString("; ");
+        step_stack.PushInstId(sem_ir.types().GetInstId(inst.element_type_id));
```
Comment (zygoloid) [inline toolchain/sem_ir/stringify_type.cpp]: We seem to be doing this a lot; maybe we should add a `PushTypeId` for this.

**Review by zygoloid — APPROVED** [review]
Looks good to me. A couple of comments but neither of them needs to be addressed in this PR.

File: toolchain/sem_ir/stringify_type.cpp (hunk 1 above)
Comment (josh11b) [reply toolchain/sem_ir/stringify_type.cpp]: Possibly, though that seems out of scope for this PR.

File: toolchain/sem_ir/stringify_type.cpp (hunk 2 above)
Comment (josh11b) [reply toolchain/sem_ir/stringify_type.cpp]: Done.

File: toolchain/sem_ir/stringify_type.cpp (hunk 1 above)
Comment (josh11b) [reply toolchain/sem_ir/stringify_type.cpp]: Oh I hadn't realized how little else was going on in this file. I've moved it out into an anonymous namespace.

**Comment by josh11b** [comment]
FYI, I did some small cleanups to the `StepStack` class.

**Review by danakj — COMMENTED** [review]
This is nice, thank you :)

**Pull request merged**